### PR TITLE
Update vendored volume visual from vispy

### PR DESCRIPTION
**Pull request by alisterburt**
# Description
This PR updates our vendored volume visual code to be in line with current vispy master. The main purpose of this PR is to make the diff between #3023 and master smaller. 

The support for alternative types of floating point textures in vispy 0.7 means that our vendored module now depends on something in a private vispy module...

```python
from vispy.visuals._scalable_textures import CPUScaledTexture3D, GPUScaledTextured3D
```

I vendored that file accordingly which also required vendoring a few other little functions, those are what you see in the vendored `vispy.gloo` package

We need to maintain our vendored version of the visual because @kevinyamauchi's changes in #3006 and the coming changes in #3023 and #2934 modify things inside the shader. 

I checked this works down to vispy 0.6.6 for the napari examples, can't go lower locally unfortunately!

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
LGTM! Maybe @kevinyamauchi can approve before merge too, should be get this in before #3037 the ray PR, or do you want to get that other one in first @kevinyamauchi? If you approve this and want to get it in first, feel free to merge in the morning if all tests are passing. Thanks @alisterburt!

**Comment by alisterburt** [comment]
thanks @sofroniewn !

I'm getting the same test errors on #3023 and can't reproduce locally (blue colormap is very much blue), not sure what's going on there

**Comment by sofroniewn** [comment]
> I'm getting the same test errors on #3023 and can't reproduce locally (blue colormap is very much blue), not sure what's going on there

hmm funny, really not sure why there would be a difference local vs CI here ... will think on it.

**Comment by tlambert03** [comment]
One common reason why things will be different between the two is that they're actually running a different set of tests (due to skips) and sometime tests have side effects (they "leak" state) that affect later tests.

Try: 
1) running locally with CI=1 pytest
2) running locally with CI=1 tox

**Comment by sofroniewn** [comment]
@alisterburt I can actually reproduce fails locally too. By default these screenshot tests are passed, but if I set `export NAPARI_POPUP_TESTS=1` and then run `pytest napari/_tests/test_with_screenshot.py` I get

```
FAILED napari/_tests/test_with_screenshot.py::test_changing_image_colormap - AssertionError:
FAILED napari/_tests/test_with_screenshot.py::test_changing_image_attenuation - assert 51 > 80
```

I havn't had a chance to look why tests fail, but that should make it easier to debug

**Comment by alisterburt** [comment]
thanks both of you! There's definitely a bug going on here. Changing the colormap in `ndisplay=2` works, switching to `ndisplay=3` correctly uses the colormap previously selected, changing colormaps in `ndisplay=3` doesn't work... looking into it

**Comment by sofroniewn** [comment]
When I run this locally I get a warning

```
GPUs can't support floating point data with more than 32-bits, precision will be lost due to downcasting to 32-bit float.
```

and 

```
WARNING: Program has unset variables: {'u_attenuation'}
```

There might be something going on with `u_attenuation`?

**Comment by alisterburt** [comment]
@sofroniewn the float64 error was recently added to vispy, I can suppress the warning in napari if you prefer? I'll look into the `u_attenuation` error, looks familiar!

**Comment by tlambert03** [comment]
I believe I handled that warning in the vispy pr that we closed, so might check there for a possible fix

**Comment by alisterburt** [comment]
will do, thanks @tlambert03

**Comment by alisterburt** [comment]
Thanks for the pointers @tlambert03 and @sofroniewn - this is now passing all tests so I think we're good to merge

**Comment by sofroniewn** [comment]
> Thanks for the pointers @tlambert03 and @sofroniewn - this is now passing all tests so I think we're good to merge

do we need to update our vispy requirement to `>=0.7.3` for this?

**Comment by alisterburt** [comment]
I don't think so, I vendored what was needed and the minimum requirements tests are passing (`vispy==0.6.4`)

**Comment by sofroniewn** [comment]
> I don't think so, I vendored what was needed and the minimum requirements tests are passing (vispy==0.6.4)

Ok nice

> @sofroniewn the float64 error was recently added to vispy, I can suppress the warning in napari if you prefer?

Yeah can you suppress this warning or fix, I still see it. Fix might be better - I guess we should just downcast ourselves if needed?

I see it when I run `python examples/nD_labels.py` for example

**Comment by alisterburt** [comment]
Added a check on layer init for `ndim == 3` and `dtype == float64`, in which case I downcast to float32, no more warnings :)

Oh no, looks like I broke some tests... can't fix right now but will get to it ASAP

**Comment by sofroniewn** [comment]
> Added a check on layer init for ndim == 3 and dtype == float64, in which case I downcast to float32, no more warnings :)

Fix works for me!

We've got what look like some unrelated test fails coming in now though - possibly connected to zarr? Could be a new upstream package that has come out?

**Comment by alisterburt** [comment]
uh oh, can't fix right now but will get to it ASAP!

File: napari/layers/image/image.py [hunk 1]
```diff
@@ -218,6 +218,10 @@ def __init__(
         else:
             ndim = len(init_shape)
 
+        # Cast 3D images in float64 to float32
+        if ndim == 3 and data.dtype == np.float64:
```
Comment (sofroniewn) [inline napari/layers/image/image.py]: > Oh no, looks like I broke some tests... can't fix right now but will get to it ASAP

Ah i see why - let's do the down cast on the vispy side we actually have this line https://github.com/napari/napari/blob/d6bc683b019c4a3a3c6e936526e29bbd59cca2f4/napari/_vispy/vispy_image_layer.py#L89 and the `fix_data_and_dtype` function here https://github.com/napari/napari/blob/d6bc683b019c4a3a3c6e936526e29bbd59cca2f4/napari/_vispy/utils_gl.py#L73 that you should be able to use. 

> uh oh, can't fix right now but will get to it ASAP!

No worries, take your time!

**Comment by alisterburt** [comment]
turns out those warnings came from the constructor in the `ImageLayerNode`, data was already downcast to float32 by the `fix_data_dtype()` function but the docstring wrongly said float64 was acceptable - fixed that and everything appears to be working now, I think this is good to go!

**Comment by sofroniewn** [comment]
> fixed that and everything appears to be working now, I think this is good to go!

Yup, I just confirmed, this all works for me, warning is now gone. I'll merge now. Thanks!!

**Pull request merged**